### PR TITLE
add demoteFunc controller option

### LINKED ISSUE #1683: Add controller options to set reconciler DemoteFunc
## Expected Behavior
Being able to set DemoteFunc when creating a controller.

## Actual Behavior
Not possible.

## Steps to Reproduce the Problem

1.
2.
3.

## Additional Info

Receive adapter internal state must be updated when demoting a bucket

### PR DESCRIPTION
**Pull request by lionelvillard**
# Changes

-  :gift: Add a controller option for configuring the reconciler demote function. 
- 
-

/kind enhancement

Fixes #1683

**Release Note**

```release-note
 :gift: Add a controller option for configuring the reconciler demote function. 
```

**Docs**

```docs

```

### CODE REVIEW

File: codegen/cmd/injection-gen/generators/reconciler_reconciler.go [hunk 1]
```diff
@@ -363,6 +363,9 @@ func NewReconciler(ctx {{.contextContext|raw}}, logger *{{.zapSugaredLogger|raw}
 			rec.skipStatusUpdates = true
 		}
 		{{- end}}
+        if opts.DemoteFunc != nil {
```
Comment (vagababov) [inline codegen/cmd/injection-gen/generators/reconciler_reconciler.go:366]: indent is off here.

Comment (lionelvillard) [reply codegen/cmd/injection-gen/generators/reconciler_reconciler.go]: space vs tab... Fixed.

**Comment by lionelvillard** [comment]
gentle ping.

**Comment by lionelvillard** [comment]
gentle ping. This PR is blocking this critical fix: https://github.com/knative/eventing/pull/4908

**Comment by vagababov** [comment]
Prow seems to be down?

cc @coryrc @chizhg

**Pull request merged**